src/rewriter/basic_rewrite_rcons.cpp

```diff
@@ -2250,6 +2259,421 @@ bool BasicRewriteRCons::ensureProofMacroQuantVarElimEq(CDProof* cdp,
+    if (negBody != negPremise)
+    {
+      Trace("brc-macro") << "- rewrite de-morgan " << negBody << " -> "
+                         << negPremise << std::endl;
+      // by de-morgan
+      tcpg.addRewriteStep(negBody,
+                          negPremise,
+                          nullptr,
+                          true,
+                          TrustId::MACRO_THEORY_REWRITE_RCONS_SIMPLE);
+    }
+    Node negRew = nm->mkNode(Kind::IMPLIES, negBody, nm->mkConst(false));
+    // F = (=> (not F) false)
+    Trace("brc-macro") << "- rewrite impl intro " << eq[0][1] << " -> "
+                       << negRew << std::endl;
+    tcpg.addRewriteStep(eq[0][1],
+                        negRew,
+                        nullptr,
+                        true,
+                        TrustId::MACRO_THEORY_REWRITE_RCONS_SIMPLE);
+    Trace("brc-macro") << "...from " << eq[0] << std::endl;
+    std::shared_ptr<ProofNode> pfn = tcpg.getProofForRewriting(eq[0]);
+    eqnorm = pfn->getResult();
+    qnorm = eqnorm[1];
+    cdp->addProof(pfn);
+    Trace("brc-macro") << "...normalized to " << qnorm << std::endl;
+  }
+  else
+  {
+    qnorm = eq[0];
+  }
+  // Now have upper set. note if all disequalities we don't care about the
+  // value of isUpper
+  std::reverse(normLits.begin(), normLits.end());
+  // make the max or min of all terms based on isUpper
+  Node iterm;
+  for (const Node& nl : normLits)
+  {
+    Node atom = nl.getKind() == Kind::NOT ? nl[0] : nl;
+    Trace("brc-macro") << "...process normalized atom " << atom << std::endl;
+    Kind k = atom.getKind();
+    Node itc = atom[1];
+    if (k != Kind::GEQ && k != Kind::LEQ)
+    {
+      itc = rewrite(nm->mkNode(
+          Kind::ADD,
+          itc,
+          nm->mkConstRealOrInt(itc.getType(), Rational(isUpper ? -1 : 1))));
+    }
+    if (iterm.isNull())
+    {
+      iterm = itc;
+    }
+    else
+    {
+      iterm = nm->mkNode(Kind::ITE,
+                         nm->mkNode(isUpper ? Kind::LT : Kind::GEQ, itc, iterm),
+                         itc,
+                         iterm);
+    }
+  }
+  Trace("brc-macro") << "Instantiation term is: " << iterm << std::endl;
+  // instantiate
+  ProofChecker* pc = d_env.getProofNodeManager()->getChecker();
+  Node iarg = nm->mkNode(Kind::SEXPR, iterm);
+  Assert(qnorm[0].getNumChildren() == 1);
+  Trace("brc-macro") << "Instantiate: " << qnorm << " / " << iarg << std::endl;
+  Node inst = pc->checkDebug(ProofRule::INSTANTIATE, {qnorm}, {iarg});
+  cdp->addStep(inst, ProofRule::INSTANTIATE, {qnorm}, {iarg});
+  Trace("brc-macro") << "Have instantiation: " << inst << std::endl;
+  Node falsen = nm->mkConst(false);
+  if (normLits.size() > 1)
+  {
+    Assert(inst.getKind() == Kind::IMPLIES && inst[1] == falsen);
+    Assert(inst[0].getKind() == Kind::AND);
+    std::vector<Node> ipremises(inst[0].begin(), inst[0].end());
+    Node currTerm = iterm;
+    // always have proven iterm >= currTerm
+    Node src;
+    size_t index = 0;
+    do
```

Comment:
I wonder if this do block would be simpler if we let the instantiation be a (negation of) sum of absolute values...

Reply:
We had discussed this before, I think it'd be roughly the same difficulty. I chose to use max terms since we already required RARE rules for them.